### PR TITLE
Simplify ScatterElements reduction codegen

### LINKED ISSUE #311: ONNX backend test suite: 181 failures across reduce/cast/resize/nllloss op families (90.1% pass)
## Summary

Running the full ONNX backend test suite against burn-onnx (via the scoreboard integration drafted in #299) reveals 181 failing tests out of 1827 CPU tests. Pass rate: **1646/1827 (90.1%)**. Runtime: 21 minutes.

Failures are concentrated in specific op families, not the integration layer itself. This issue tracks the correctness gaps that need to be addressed before submitting to the ONNX Backend Scoreboard.

## Test Environment

- Branch: `issue/299-onnx-backend-scoreboard`
- Backend wrapper: `scoreboard/backend.py`
- Runner: `scoreboard/runner/` (burn + burn-ndarray, NdArray<f32> backend)
- onnx2burn flags: `--no-development --no-simplify`
- Full raw log: https://gist.github.com/antimora/2178ee7beddfe57d5cda5f3edc5aeb63
- Full failing test list: https://gist.github.com/antimora/a0ee7679550aa632c671ef40412e5359

## Failures by Op Family

| Op Family | Count | Root Cause (preliminary) |
|---|---:|---|
| `reduce_sum` | 24 | Output shape differs (keepdims/rank handling) |
| `castlike` | 18 | Exotic dtypes: FLOAT4E2M1, FLOAT8E*, INT2/4, UINT2/4 |
| `reduce_l1` | 14 | Output shape (keepdims, empty set) |
| `reduce_l2` | 14 | Output shape (keepdims, empty set) |
| `nllloss` | 12 | Weight handling, reduction modes |
| `resize` | 11 | Algorithm differences (cubic, nearest modes) |
| `cast` | 10 | Same exotic dtypes as castlike |
| `reduce_log_sum` | 9 | Axis handling, empty set |
| `attention` | 8 | Numerical precision on 3D/4D with causal/bias/mask |
| `reduce_max` | 7 | Output shape + bool inputs |
| `reduce_min` | 7 | Output shape + bool inputs |
| `reduce_prod` | 7 | Output shape (keepdims) |
| `reduce_mean` | 6 | Output shape (keepdims) |
| `triu` | 6 | Negative diagonal, edge cases |
| `tril` | 5 | Negative diagonal, edge cases |
| `averagepool` | 4 | `ceil_mode` with same padding |
| `bernoulli` | 4 | Random op (may need seed passthrough) |
| `bitwise` | 3 | uint64 support |
| `bitshift` | 2 | uint64 support |
| `maxpool` | 1 | `ceil_mode` output size |
| `det` | 1 | Determinant op (possibly unsupported) |
| `gelu` | 1 | Tanh approximation precision |
| `mod` | 1 | uint64 support |
| `add`, `sub`, `mul`, `div`, `max`, `min` | 6 | uint64 dtype support |

Total: 181

## Themes

Looking across failures, three dominant root causes emerge:

### 1. Reduce ops (~85 failures)
Almost half of all failures are in reduce ops (`reduce_sum`, `reduce_l1`, `reduce_l2`, `reduce_log_sum`, `reduce_max`, `reduce_min`, `reduce_prod`, `reduce_mean`, `reduce_sum_square`). The common error pattern is `DESIRED: array([3, 2])` - meaning the output shape differs from what ONNX expects. This suggests a systemic issue with:
- `keepdims` handling (especially `do_not_keepdims`)
- `noop_with_empty_axes` handling (empty set tests)
- Bool input support (`reduce_max_bool_inputs`, `reduce_min_bool_inputs`)
- Negative axis normalization

### 2. uint64 support (~12 failures)
Many elementwise ops (`add`, `sub`, `mul`, `div`, `max`, `min`, `mod`, `bitwise`, `bitshift`) fail with `DESIRED: dtype('uint64')`. Burn's NdArray backend doesn't expose u64 as a tensor element type through the standard float/int paths.

### 3. Exotic float/int dtypes (~28 failures)
`cast` and `castlike` tests fail for FLOAT4E2M1, FLOAT8E4M3FN, FLOAT8E4M3FNUZ, FLOAT8E5M2, FLOAT8E5M2FNUZ, INT2, INT4, UINT2, UINT4. These are sub-byte and low-precision float types that Burn doesn't currently support. Likely low-priority for initial scoreboard submission.

## Sample Errors

```
test_reduce_sum_do_not_keepdims_example_cpu:  DESIRED: array([3, 2])
test_reduce_max_bool_inputs_cpu:               [False]] != [[ True]]
test_add_uint64_cpu:                           DESIRED: dtype('uint64')
test_cast_FLOAT4E2M1_to_FLOAT_cpu:             DESIRED: array([3, 5])
test_tril_neg_cpu:                             [6, 7, 7, 0, 0]])
test_attention_3d_causal_cpu:                  0.896761, 0.406733, 0.552078, ...
test_bernoulli_double_cpu:                    DESIRED: array([0., 1., 1., 0., 0., ...])
test_resize_downsample_scales_cubic_align_corners_cpu:  [12.161532, 13.556724, ...]
```

## Suggested Priority

**P0 - Systemic fixes with high test yield:**
1. Investigate reduce ops shape/keepdims bug (unlocks ~85 tests)
2. Review Tril/Triu negative diagonal handling (unlocks 11 tests)

**P1 - Algorithm correctness:**
3. Resize cubic and nearest mode edge cases (11 tests)
4. AveragePool/MaxPool ceil_mode (5 tests)
5. NLLLoss weight and reduction handling (12 tests)
6. Attention numerical precision (8 tests)

**P2 - Type system expansion:**
7. uint64 support through tensor ops (~12 tests)
8. Exotic dtype support for cast/castlike (~28 tests) - can be skipped for initial submission

**P3 - Inherently noisy:**
9. Bernoulli (4 tests) - stochastic, may need special handling or exclusion

## Reproducing

```bash
git checkout issue/299-onnx-backend-scoreboard
cargo build --release -p burn-onnx --bin onnx2burn
cd scoreboard/runner && cargo build && cd ..
uv pip install onnx pytest numpy
ONNX2BURN="$(pwd)/../target/release/onnx2burn" uv run python test_local.py -k "not _cuda" -v
```

See `scoreboard/README.md` for full setup details.

---

Related: #299 (scoreboard integration prep)

### LINKED ISSUE #486: Update burn-onnx scatter codegen to use expanded IndexingUpdateOp support
### Feature description

Update the `burn-onnx` codebase to take advantage of the expanded `IndexingUpdateOp` support added in `burn` so ONNX scatter operators can use native update modes instead of workarounds and generated element-by-element fallback loops.

This follow-up should update the `ScatterElements` and `ScatterND` handling to map ONNX reduction modes directly onto the corresponding Burn scatter/update operations:

| ONNX reduction | Target Burn op |
|---|---|
| `none` | `Assign` |
| `add` | `Add` |
| `mul` | `Mul` |
| `max` | `Max` |
| `min` | `Min` |

The current codebase still relies on:
- a gather-diff-scatter workaround for `none`
- generated per-element fallback loops for `mul`, `max`, and `min`

After the backend support landed in `burn`, these paths should be simplified to use direct `scatter()` / related update operations wherever possible.

### Feature motivation

This keeps `burn-onnx` aligned with the new capabilities added in `burn` and reduces generated code complexity for ONNX models using `ScatterElements` or `ScatterND`.

Expected benefits include:
- simpler code generation
- less generated boilerplate per model
- better readability and maintainability
- closer alignment with ONNX reduction semantics
- removal of workaround logic that is no longer necessary

### (Optional) Suggest a Solution

- Audit the current `ScatterElements` implementation and replace the `none` reduction workaround with direct assignment-based scatter support.
- Replace element-by-element fallback logic for `mul`, `max`, and `min` with direct scatter/update calls using the new `IndexingUpdateOp` variants.
- Apply the same cleanup to `ScatterND`, which supports the same reduction modes.
- Add or update tests covering all five ONNX reduction modes for both operators.
- Reference the upstream `burn` work that introduced `Assign`, `Mul`, `Max`, and `Min` support in `IndexingUpdateOp`.

Related upstream context:
- burn issue: https://github.com/tracel-ai/burn/issues/4477
- ONNX follow-up context: https://github.com/tracel-ai/burn-onnx/issues/125

### PR DESCRIPTION
**Pull request by antimora**
### Checklist

- [x] Confirmed that `cargo xtask validate` command has been executed. Format and lint pass. Its workspace test stage fails locally on `raspberry-pi-pico` only (`rp-pac` hits `invalid Mach-O section specifier` on darwin, and `ssmarshal` no longer compiles against current `serde`) - both pre-existing and excluded by `CI_EXCLUDE`. `cargo test --workspace` with CI's exclusion list passes.
- [x] Confirm relevant documentation has been updated.
- [x] For new/modified ONNX operators: verified implementation matches `onnx-spec/ops/<OpName>.md`.

### Related Issues/PRs

- Closes #486
- Closes #311 (`test_scatter_elements_with_negative_indices` promoted from `fail-compare` to `pass`)
- Upstream backend gap filed as tracel-ai/burn#5522
- Follow-up to tracel-ai/burn#4477, which extended `IndexingUpdateOp` and threaded it through `scatter` / `select_assign` but explicitly left the per-backend implementations out

### Changes

`ScatterND` already lowered all five ONNX reductions natively and needed no change. `ScatterElements` did, and the reason it could not simply call `scatter()` is that element-wise `scatter` is only partly implemented:

| Backend | Assign | Add | Mul | Min | Max |
|---|---|---|---|---|---|
| ndarray | yes | yes | yes | no | no |
| flex (default) | no | yes | yes | no | no |
| cubecl | no | yes | yes (not at the pinned rev) | no | no |
| tch | no | yes | yes | no | no |

`burn-backend`'s default `float_scatter` forwards only `Add` and `unimplemented!()`s the rest, so a backend that does not override it inherits Add-only support and panics at runtime rather than failing to compile.

`scatter_nd`, by contrast, implements all five on every backend. ScatterElements writes to

```
data[p_0, .., indices[p], .., p_{r-1}]   for every p in the index shape
```

so materializing those coordinates as index tuples turns it into a ScatterND and every reduction becomes a native backend op:

- **`none` / `mul` / `min` / `max`** lower to `scatter_nd(coords, updates, IndexingUpdateOp::{Assign,Mul,Min,Max})`.
- **`add`** keeps element-wise `scatter`, the one update op every backend implements and the only one ONNX defines for duplicate indices.
- **bool `none`** round-trips through i64, matching `scatter_nd.rs`; bool scatter only maps to a logical `or`, which cannot clear a bit.

The non-axis coordinate columns come from a flat `arange` decomposed by row-major strides, so the lowering stays vectorized:

```rust
se_flat.clone()
    .div_scalar(se_strides[se_d] as i64)
    .remainder_scalar(se_idx_dims[se_d] as i64)
    .reshape([se_n, 1])
```

The fallback this replaces called `indices.to_data()` - a device-to-host sync and readback - and then issued a `narrow` + `slice_assign` pair per scattered element. Generated code is ~25 lines against the previous 35, with no readback and no per-element kernel launches.

**Negative indices** are now normalized on every path, including `add`, which never did it. ONNX allows values down to `-dim_size`; the removed loop folded them but the `none` and `add` paths did not, which is what #311 was tracking.

#### On the first commit

The first commit expressed `none`/`min`/`max` as a scatter-**add** of a delta (`e + max(0, u - e) = max(e, u)`). Review caught three defects, all of which are now covered by tests that failed before the fix:

- **Infinities.** `max` with `data = -inf` and `update = 0.5` computes `-inf + inf` and yields NaN. Assignment expressed as an add is unrecoverable when the existing value is infinite, no matter how the delta is clamped. Narrow integer types have the same problem in the intermediate subtraction.
- **Duplicate indices.** Each delta was computed from the original gathered value and then summed, so max over `[2, 3]` into a zeroed target gave 5 instead of 3.
- **Negative indices**, as above.

The `scatter_nd` lowering fixes all three, and also removes the GPU regression the first commit would have introduced for `mul` (cubecl has no element-wise `scatter` Mul at the pinned burn rev, but does implement `scatter_nd` Mul).

### Testing

- Snapshots updated for every codegen branch.
- New integration tests, each failing before the fix: negative indices on `none`/`max`/`mul`, `max` with `-inf`, `max` with duplicate indices, and a rank-3 `axis=1` case exercising both a non-unit outer stride and the unit inner stride.
- The existing `max`/`min` tests were weak - every update lost to the existing value, so the "update wins" branch never ran. Widened to mixed inputs, expectations taken from `onnx.reference.ReferenceEvaluator`.
- Added `scatter_elements_bool` covering the bool path with targets that are both set and cleared.
- Full runs: 1211 burn-onnx, onnx-tests, onnx-ir, and 835 onnx-official-tests (up from 834). `cargo test --workspace` with CI's exclusion list exits 0 across 35 test binaries.

### Known limitation

Burn documents duplicate indices as undefined for `scatter_nd` with `Assign`/`Mul`/`Min`/`Max`, while ONNX defines them. In practice the flex and ndarray implementations fold duplicates sequentially. This is upstream of this PR.

tracel-ai/burn#5522 is now a performance item rather than a correctness one for this operator: once element-wise `scatter` gains the missing update ops, `none`/`mul`/`min`/`max` can drop the coordinate construction and call `scatter()` directly.

### CODE REVIEW

File: crates/burn-onnx/src/import/burn/node/scatter_elements.rs [hunk 1]
```diff
@@ -24,114 +24,56 @@ impl NodeCodegen for onnx_ir::scatter_elements::ScatterElementsNode {
...
+            // assignment round-trips through i64 and uses the same delta identity as below.
+            return quote! {
                 let #output = {
-                    let gathered = #data.clone().gather(#dim, #indices.clone());
-                    #data.scatter(#dim, #indices, #updates - gathered, burn::tensor::IndexingUpdateOp::Add)
+                    let bool_data = #data.int().cast(burn::tensor::DType::I64);
+                    let bool_updates = #updates.int().cast(burn::tensor::DType::I64);
+                    let gathered = bool_data.clone().gather(#dim, #indices.clone());
+                    bool_data
+                        .scatter(#dim, #indices, bool_updates - gathered, burn::tensor::IndexingUpdateOp::Add)
```
[the remark this answers is not in the document]
Comment (antimora) [reply crates/burn-onnx/src/import/burn/node/scatter_elements.rs]: Valid, and fixed in 12c6be78.

Confirmed the regression: the removed loop did `if idx < 0 { idx += dim_size; }`, so `mul`/`max`/`min`/bool-`none` handled negatives before this PR and stopped doing so. The numeric `none` and `add` paths never normalized at all, which is why `test_scatter_elements_with_negative_indices` was sitting in `expectations.toml` as `fail-compare` under #311.

Normalization is now a shared prologue on every path, so `add` gets it too:

```rust
let se_axis_size = se_data.dims()[axis] as i64;
let se_indices = indices.cast(DType::I64);
let se_negative = se_indices.clone().lower_elem(0i64);
let se_corrected = se_indices.clone() + se_axis_size;
let se_indices = se_indices.mask_where(se_negative, se_corrected);
```

Only one scalar dim size is involved because the scatter axis is fixed, so this is cheaper than the `[1, ..., 1, K]` broadcast `indexing_helpers::negative_index_normalize` needs for ScatterND.

Added `scatter_elements_negative_indices`, `scatter_elements_max_negative_indices` and `scatter_elements_mul_negative_indices` (all failed before the fix), plus a rank-3 case mixing negative and positive indices. `test_scatter_elements_with_negative_indices` now passes and is promoted to `pass`, so the official suite goes 834 -> 835.

**Comment by antimora** [comment]
The review's four suppressed comments were all correct. They don't exist as threads to reply into, so here is each one and what changed. All of it is fixed in 12c6be78, which replaces the delta form with a `scatter_nd` lowering.

I wrote each as a failing test first; all five failed before the fix and pass after.

---

> These delta formulas are incorrect when indices repeat. [...] for `data=0`, duplicate max updates `[2, 3]` produce `5` instead of ONNX's `3`

Correct. Each delta was computed from the original gathered value and then summed by the scatter-add. `scatter_elements_max_with_duplicate_indices` covers it.

> The arithmetic identity is not valid for all ONNX numeric values supported here. For example, max with `data=-inf` and `update=1` computes `-inf + inf = NaN`

Correct, and this was the one that settled the design. Assignment expressed as an add is unrecoverable when the existing value is infinite: `e + (max(e,u) - e)` is NaN for `e = -inf` regardless of how the delta is clamped. Narrow integer types have the matching problem in the intermediate subtraction. `scatter_elements_max_with_infinity` covers it.

> This direct `Mul` lowering regresses valid models on the repository's pinned Burn revision: the WGPU/Metal/CUDA paths still reach the backend `unimplemented!()`

Correct. This is also fixed, though not the way the comment suggests: `mul` now goes through `scatter_nd`, which cubecl does implement for all five ops at the pinned rev, so the GPU regression is gone without waiting on a Burn bump.

> Burn documents duplicate targets as undefined for `IndexingUpdateOp::Mul`, but ONNX requires multiplying every update into the target

Right about the mismatch. Routing `mul` through `scatter_nd` doesn't make Burn's documented contract stronger, but it does hand the reduction to the backend instead of encoding it as arithmetic, and the flex and ndarray `scatter_nd` implementations fold duplicates sequentially. `add` stays on element-wise `scatter`, since it is the one op both Burn and ONNX define for duplicates.

---

**The fix.** `scatter_nd` implements `Assign`/`Add`/`Mul`/`Min`/`Max` on every backend, unlike element-wise `scatter`. ScatterElements writes to `data[p_0, .., indices[p], .., p_{r-1}]` for every `p` in the index shape, so materializing those coordinates as index tuples turns it into a ScatterND and every reduction becomes native.

The non-axis coordinate columns are recovered from a flat `arange` decomposed by row-major strides, so nothing here reintroduces the host readback the original PR removed:

```rust
se_flat.clone()
    .div_scalar(se_strides[se_d] as i64)
    .remainder_scalar(se_idx_dims[se_d] as i64)
    .reshape([se_n, 1])
```

Generated code is ~25 lines against the original 35-line loop, but with no `to_data()` sync and no per-element kernel launches.

**Coverage.** New tests for negative indices (three reductions), infinities, duplicate indices, and a rank-3 `axis=1` case that exercises both a non-unit outer stride and the unit inner stride. `test_scatter_elements_with_negative_indices` is promoted from `fail-compare` to `pass`, closing #311; the official suite goes 834 -> 835.

**Still open, and now the only thing tracel-ai/burn#5522 blocks:** ScatterElements no longer needs `Assign`/`Min`/`Max` on element-wise `scatter` to be correct, so that issue is a performance item rather than a correctness one. Duplicate-index semantics for `Mul`/`Min`/`Max` remain undefined per Burn's `scatter_nd` docs; ONNX defines them, so that gap is real but is upstream of this PR.
